src/main/java/io/cryostat/rules/MatchExpressionManager.java

```diff
@@ -0,0 +1,211 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package io.cryostat.rules;
+
+import java.lang.reflect.Type;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import javax.script.ScriptException;
+
+import io.cryostat.core.log.Logger;
+import io.cryostat.platform.PlatformClient;
+import io.cryostat.platform.ServiceRef;
+
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import dagger.Lazy;
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+
+public class MatchExpressionManager {
+    private final MatchExpressionValidator matchExpressionValidator;
+    private final Lazy<MatchExpressionEvaluator> matchExpressionEvaluator;
+    private final PlatformClient platformClient;
+    private final MatchExpressionDao dao;
+    private final Gson gson;
+    private final Logger logger;
+
+    MatchExpressionManager(
+            MatchExpressionValidator matchExpressionValidator,
+            Lazy<MatchExpressionEvaluator> matchExpressionEvaluator,
+            PlatformClient platformClient,
+            MatchExpressionDao dao,
+            Gson gson,
+            Logger logger) {
+        this.matchExpressionValidator = matchExpressionValidator;
+        this.matchExpressionEvaluator = matchExpressionEvaluator;
+        this.platformClient = platformClient;
+        this.dao = dao;
+        this.gson = gson;
+        this.logger = logger;
+    }
+
+    public int addMatchExpression(String matchExpression)
+            throws MatchExpressionValidationException {
+        matchExpressionValidator.validate(matchExpression);
+        MatchExpression expression = dao.save(new MatchExpression(matchExpression));
+        return expression.getId();
+    }
+
+    public Optional<MatchExpression> get(int id) {
+        return dao.get(id);
+    }
+
+    public List<MatchExpression> getAll() {
+        return dao.getAll();
+    }
+
+    public boolean delete(int id) {
+        return dao.delete(id);
+    }
+
+    public Set<ServiceRef> resolveMatchingTargets(int id) {
+        Optional<MatchExpression> matchExpression = dao.get(id);
+        if (matchExpression.isEmpty()) {
+            return Set.of();
+        }
+        return resolveMatchingTargets(matchExpression.get());
+    }
+
+    public Set<ServiceRef> resolveMatchingTargets(MatchExpression expr) {
+        Set<ServiceRef> matchedTargets = new HashSet<>();
+        for (ServiceRef target : platformClient.listDiscoverableServices()) {
+            try {
+                if (matchExpressionEvaluator.get().applies(expr.getMatchExpression(), target)) {
```

Comment:
This looks like it can be refactored in terms of `resolveMatchingTargets(String expr, Collection<ServiceRef> targets)` I think

Reply:
Or vice-versa.

Basically just the filter function can be either set to that `contains` check or the simple function `t -> true` to pass everything.